⬆️ UPDATE: Siesta MaX-1.1

albgar opened the pull request:
The updated role builds and deploys version 'MaX-1.1' of Siesta (with support for pseudopotentials in PSML format, and for Lua scripting, among other improvements). It does not build support for the ELSI library as this is more of a performance feature.

  - It installs the siesta executable, as well as several post-processing utilities, in /usr/local/bin. 
  - Requisite domain-specific libraries (libpsml, xmlf90, libgridxc, flook) are in /usr/local/lib, and their module files in /usr/local/include.
- It uses ldconf to add to the shared-library search path.
- It installs files for several pseudopotential families in /usr/local/share/siesta/psml-files.
- For the interface with the AiiDA role, it defines the `siesta_executables` variable.

ltalirz reviewed (dismissed):
Thanks a lot @albgar, I can see this was quite a lot of work! 

Just out of curiosity, since we were discussing this previously - have there been any developments on the spack/easybuild/conda front for siesta builds?

For now it's perfectly fine to install packages from source here in the ansible role, but of course this all needs to be maintained, i.e. when the next version of siesta is released, it likely won't be as easy as just increasing the siesta version in the `default.yaml` ;-)

ltalirz commented:
@chrisjsewell Feel free to merge this or to review as well

chrisjsewell commented on defaults/main.yml:
```diff
@@ -8,118 +25,161 @@ siesta_data_folder: "{{ siesta_prefix }}/share/siesta"
-siesta_xmlf90_root: ""
-siesta_psml_root: ""
-siesta_gridxc_root: ""
+xmlf90_version: "1.5.4"
```
Can you add the `siesta_` prefix to all these variables please.
When building Quantum Mobile obviously we have a lot of roles and variables, so it helps to keep track of where they came from.

chrisjsewell commented on tasks/flook.yml:
```diff
@@ -0,0 +1,45 @@
+  get_url:
+    url: "{{ flook_url }}"
+    dest: "{{ flook_code_folder }}/{{ flook_src_archive }}"
+  register: flook_download
```
add some retries in case of bad connections

```suggestion
  register: flook_download
  until: not flook_download.failed
  retries: 3
  delay: 10
```

chrisjsewell commented on tasks/gridxc-multi.yml:
```diff
@@ -0,0 +1,94 @@
+  get_url:
+    url: "{{ libgridxc_url }}"
+    dest: "{{ libgridxc_code_folder }}/{{ libgridxc_src_archive }}"
+  register: libgridxc_download
```
as above:

```suggestion
  register: libgridxc_download
  until: not libgridxc_download.failed
  retries: 3
  delay: 10
```

chrisjsewell commented on tasks/flook.yml:
```diff
@@ -0,0 +1,45 @@
+    creates: "{{ flook_topdir }}/libflookall.a"
+  register: flook_make
+
+- name: Install flook lib
+  command: cp -p libflookall.a "{{ flook_prefix }}/lib"
+  args:
+    chdir: "{{ flook_topdir }}"
+    creates: "{{ flook_prefix }}/lib/libflookall.a"
+  become: yes
+  register: flook_install_lib
+
+- name: Install flook module
+  command: cp -p flook.mod "{{ flook_prefix }}/include"
+  args:
+    chdir: "{{ flook_topdir }}"
+    creates: "{{ flook_prefix }}/include/flook.mod"
+  become: yes
+  register: flook_install_mod
```
Can you not just use https://docs.ansible.com/ansible/2.9/modules/copy_module.html for these?

chrisjsewell commented on tasks/gridxc-multi.yml:
```diff
@@ -0,0 +1,94 @@
+  args:
+    chdir: "{{ libgridxc_topdir }}/build_dp"
+    creates: "{{ libgridxc_topdir }}/build_dp/Makefile"
+  register: libgridxc_dp_configure
```
I'm noticing your registering a variable for every task. I don't think this is necessary unless you are actually going to use them

chrisjsewell commented on tasks/gridxc-multi.yml:
```diff
@@ -0,0 +1,94 @@
+
+# --------- mpi_dp version
+- name: Configure libgridxc mpi_dp
+  shell: "MPIFC=mpif90 MPICC=mpicc ../configure --enable-multiconfig --with-mpi {{ libgridxc_libxc_incantation }} --prefix={{ libgridxc_prefix }}"
```
you could just use

```suggestion
  shell: "../configure --enable-multiconfig --with-mpi {{ libgridxc_libxc_incantation }} --prefix={{ libgridxc_prefix }}"
  environment:
  	MPIFC: mpif90
  	MPICC: mpicc
```

but not strictly necessary

chrisjsewell commented on tasks/libpsml.yml:
```diff
@@ -0,0 +1,34 @@
+  get_url:
+    url: "{{ libpsml_url }}"
+    dest: "{{ libpsml_code_folder }}/{{ libpsml_src_archive }}"
+  register: libpsml_download
```
```suggestion
  register: libpsml_download
  until: not libpsml_download.failed
  retries: 3
  delay: 10
```

chrisjsewell commented on tasks/psml-pps.yml:
```diff
@@ -0,0 +1,35 @@
+  get_url:
+    url: "{{ psml_pps_url }}"
+    dest: "{{ psml_pps_code_folder }}/{{ psml_pps_src_archive }}"
+  register: psml_pps_download
```
```suggestion
  register: psml_pps_download
  until: not psml_pps_download.failed
  retries: 3
  delay: 10
```

chrisjsewell commented on tasks/siesta.yml:
```diff
@@ -0,0 +1,88 @@
+  get_url:
+    url: "{{ siesta_url }}"
+    dest: "{{ siesta_code_folder }}/{{ siesta_src_archive }}"
+  register: siesta_download
```
```suggestion
  register: siesta_download
  until: not siesta_download.failed
  retries: 3
  delay: 10
```

chrisjsewell commented on tasks/main.yml:
```diff
@@ -1,162 +1,65 @@
-    mode: '0755'
-  register: eigfat2plot_install
+# Compile and install xmlf90
+- include: xmlf90.yml
```
include is being deprecated: https://docs.ansible.com/ansible/2.9/modules/include_module.html

```suggestion
- include-tasks: xmlf90.yml
```

chrisjsewell replied:
and the same for the others below

chrisjsewell reviewed (changes requested):
thanks @albgar, just some nitpicks

chrisjsewell commented:
@albgar on a related matter, I've actually this week created a role for building abinit.
In this I had an issue building for  Ubuntu 20.04, because `libblacs-mpi-dev` was not available on apt (see apt installs here https://github.com/marvel-nccr/ansible-role-abinit/blob/master/tasks/main.yml).
But in https://github.com/siesta-project/ansible-role-siesta/blob/master/tasks/main.yml I notice that you just skip installing it for ubuntu > 16.
So is it just not required for these distros, or is there another package that takes its place?

albgar commented:
> @albgar on a related matter that you may be able to help with, I've actually this week created a role for building abinit.
> In this I had an issue building for Ubuntu 20.04, because `libblacs-mpi-dev` was not available on apt (see apt installs here https://github.com/marvel-nccr/ansible-role-abinit/blob/master/tasks/main.yml, also link to docs https://www.netlib.org/blacs/).
> But in https://github.com/siesta-project/ansible-role-siesta/blob/master/tasks/main.yml I notice that you just skip installing it for ubuntu > 16.
> So is it just not required for these distros, or is there another package that takes its place?

Newer versions of scalapack have the blacs "built-in". So you just need to install the scalapack-dev (I forgot the exact name) package.

chrisjsewell commented:
thanks! ok I'll try that

albgar commented:
@chrisjsewell: I have addressed all your comments (I think) and pushed, and the tests pass. But I do not really know how to tell the Github UI that I have done the checks requested...  
I hope you can override this, or tell me what I need to do.

chrisjsewell commented:
> But I do not really know how to tell the Github UI that I have done the checks requested...

Thanks @albgar, I think what you are looking for is that you just click on the circle here, to request my re-review:

![image](https://user-images.githubusercontent.com/2997570/100528890-ed165b00-31e1-11eb-8a85-bd0e76cd78f6.png)

Also, in case you didn't know, you can use `pre-commit install` and this will ensure these tests pass before a commit is made.

chrisjsewell commented:
> The updated role builds and deploys version 'MaX-1.1' of Siesta

Also one last thing, similar to what @ltalirz asked, with the changes to this role, what versions does it now support; only MaX-1.1 or is it e.g. back-compatible to v4.1, and what will potentially be involved to "upgrade" to the next version of siesta.

Also in terms of aiida-siesta, I assume the current version of that supports this version of siesta (and since what aiida-siesta version).

Perhaps some information like this could be added to the README.md, and also your comment about not supporting the ELSI library etc

albgar commented:
We have currently a number of features in development waiting to be merged, and at the same time we are in the middle of a re-design of the building system for Siesta, so things are in flux. Hopefully, with the new building system in place and a more sedate version evolution, the role can inherit a bit more stability.

I have added a few more comments to README.md.

chrisjsewell commented:
Thanks again @albgar but one last thing 😬 

So when adding the new Abinit role to Quantum Mobile, I found that all the libxc tests were failing when using the apt installed `libxc-dev`, which installs v4.0.0 (see https://github.com/marvel-nccr/ansible-role-abinit/issues/9). But when I directly compiled v4.3.4 and linked to that it all worked.
So I've created a seperate https://github.com/marvel-nccr/ansible-role-libxc role and in https://github.com/marvel-nccr/ansible-role-abinit/pull/10 allow for the direct specification of the path to libxc, e.g. pointing to `/usr/local/libxc-4.3.4` rather than the apt installed `/usr/local`.

Given that Siesta also uses libxc;
(a) will one of the default tests identify any issues with libxc, i.e. use a PAW xc-functional, and
(b) if there are any issues, how easy/hard would it be to allow for a similar direct specification of the libxc path

albgar commented:
This is strange... in the virtual machine I build on focal with my current role the version of libxc-dev is 4.3.4-1 (also stated in https://packages.ubuntu.com/search?keywords=libxc-dev&searchon=names&suite=focal&section=all), and not 4.0.0 as you seem to imply.

Nevertheless, the idea to install libxc explicitly is good, as we would have full control over it. 
However, I would like to re-use the Siesta role for things other than the vanilla Quantum Mobile, and then it would be better if the installation of libxc is local to the role, instead of assuming that somehow we are installing libxc earlier in the sequence of tasks.  On the other hand, libxc installation can be time consuming, and the tests take long, so i can see the benefits of a "canonical" installation for everybody.

chrisjsewell commented:
Yep that was the idea; to keep it modular. 
The abinit role does not repent on the libxc see one, it simply defaults to apt installing unless you set a predefined path.
So this would not preclude the use of abinit or siesta outside of QM.

I note the link you provide only shows Ubuntu 20.04, perhaps for Ubuntu 18.04 (i.e. as currently used by QM) it is not as up-to-date)?

chrisjsewell commented:
Yep indeed it's a different version for 18.04: https://packages.ubuntu.com/search?keywords=libxc-dev&searchon=names&suite=bionic&section=all

albgar commented:
I have implemented a mechanism to use the libxc role if the distribution's libxc-dev package is not up to standard. So in this sense I am using the libxc role as a "subroutine". However, a few things would need to be improved to make this more robust (some probably reflect my limited ansible skills):

- Since compilation of libxc is expensive, before "calling" the libxc role I check for existence of the /usr/local/libxc-x.y.z directory, where x.y.z is the version I want. This is "hardwired" to the current installation prefix in the libxc role. If it changes, it will break. It would be great if the role could "return" quickly when called with a version that is already installed, and provided the installation path to the client.

- The libxc role does not currently install shared libraries. A client then needs to disable shared library support (e.g. in libgridxc) to use it. A role variable could be used to request shared libraries. However, this introduces more complications: what would happen if another role further down on the QM installation list actually installs a libxc-dev package, which has precedence in the shared libs search path?

Regarding the installation path, I see it as a welcome development that you are using sub-directories in /usr/local. Could everybody else do so? It would be good to have things in /usr/local/siesta-x.y.z, for example, in case we want to install different versions. Actually, support for modules (of the lmod kind) might be good too.

albgar commented:
The CI checks fail because the libxc role cannot be found. How do I fix this?

chrisjsewell commented:
If you have a look on the abinit role's molecule folder, you can see ho I do it there: look at the molecule.yml and requirements.yml

chrisjsewell commented on tasks/libxc.yml:
```diff
@@ -0,0 +1,50 @@
+  # We have to agree on this!!
+  - name: Define variable
+    set_fact:
+      siesta_libxc_root: "/usr/local/libxc-{{ siesta_libxc_version }}"
```
can this not just go in defaults/main.yml? (just below siesta_libxc_version)

albgar replied:
Ok. I was deferring this to libxc.yml in case the libxc role got smarter along the lines suggested in a previous comment, but we can assume that the libxc role has the installation path hardwired for now.

chrisjsewell commented on molecule/default/molecule.yml:
```diff
@@ -16,7 +16,20 @@ platforms:
+        vars:
+          ansible_python_interpreter: /usr/bin/python3
+          run_tests: true
+          test_libxc: "${MOLECULE_LIBXC:-no}"  # run marvel-nccr.libxc first
```
```suggestion
```

this is not used in this role (copy/paste from abinit)

chrisjsewell commented on defaults/main.yml:
```diff
@@ -8,118 +43,161 @@ siesta_data_folder: "{{ siesta_prefix }}/share/siesta"
+siesta_xmlf90_src_archive: "{{ siesta_xmlf90_src }}.tar.gz"
+siesta_xmlf90_code_folder: "/tmp"
+siesta_xmlf90_prefix: "/usr/local"
+siesta_xmlf90_url: "https://launchpad.net/xmlf90/trunk/1.5/+download/xmlf90-1.5.4.tar.gz"
```
I note that here and for `siesta_libpsml_url`, the URLs do not "dynamically" change when setting `siesta_xmlf90_version`.
One might expect that you could just set the version number and then that version would be installed, when in fact you need to change both the version number and URL.
However, I see that it is not as simple as `"https://launchpad.net/xmlf90/trunk/1.5/+download/xmlf90-{{ siesta_xmlf90_version }}.tar.gz"` because this does not account for the `1.5` in the path. So I'm not going to insist that you change this.

albgar replied:
The underlying issue here is that the packages you can get directly from Gitlab using the tag are "raw", and need an extra step of "autoreconf".  The tarballs that the Siesta role uses now are special, in that they are pre-autoreconf'ed, and uploaded into Gitlab as blobs. They can be downloaded with the reference hash that you see.  I am going to try whether autoreconf works.

albgar replied:
I have now implemented this. While doing so I discovered a minor issue in xmlf90. It has been fixed and the version upped to 1.5.5.

chrisjsewell commented on defaults/main.yml:
```diff
@@ -8,118 +43,161 @@ siesta_data_folder: "{{ siesta_prefix }}/share/siesta"
-siesta_psml_root: ""
-siesta_gridxc_root: ""
+siesta_xmlf90_version: "1.5.4"
+siesta_xmlf90_minor_version: "1.5.4"
```
I see these `_minor_version` variables set for all the packages, but they are all the same as the `_version` variable and not actually used anywhere.
So are they really necessary?

chrisjsewell commented:
> The libxc role does not currently install shared libraries. A client then needs to disable shared library support (e.g. in libgridxc) to use it. A role variable could be used to request shared libraries. However, this introduces more complications: what would happen if another role further down on the QM installation list actually installs a libxc-dev package, which has precedence in the shared libs search path?

Yes I must admit, I'm not an expert in the use of shared libraries.
I did have a brief look before, if there was anything similar to python/conda virtual environments in the "apt world", but I didn't see anything obvious (although now that you mention lmod I guess that is somewhat similar).
Any good reading/resources you could point me to would be welcome.  

> Regarding the installation path, I see it as a welcome development that you are using sub-directories in /usr/local. Could everybody else do so? It would be good to have things in /usr/local/siesta-x.y.z, for example, in case we want to install different versions. Actually, support for modules (of the lmod kind) might be good too.

Obviously the drawback is that the executables will not then be in the `$PATH` by default, and its just not something we have needed for now in Quantum Mobile, where we only have a single version of each code, and don't necessarily require the extra complication.

But yes I could see lmod as useful for the more general use of these roles.

Perhaps using something like https://github.com/idiv-biodiversity/ansible-role-lmod
Then each role could write a ModuleFile and also, in some way, contribute to a [meta module](https://lmod.readthedocs.io/en/latest/015_writing_modules.html#meta-modules) that just loads the "default" versions of all codes.

chrisjsewell reviewed (approved):
Thanks for all the work!
